src/pure/RealTimeQueue.mo

```diff
@@ -0,0 +1,1026 @@
+/// Double-ended immutable queue with guaranteed `O(1)` push/pop operations (caveat: high constant factor).
+/// For a default immutable queue implementation, see `pure/Queue`.
+///
+/// This module provides an alternative implementation with better worst-case performance for single operations, e.g. `pushBack` and `popFront`.
+/// These operations are always constant time, `O(1)`, which eliminates spikes in performance of `pure/Queue` operations
+/// that are caused by the amortized nature of the `pure/Queue` implementation, which can lead to `O(n)` worst-case performance for a single operation.
+/// The spikes in performance can cause a single message to take multiple more rounds to complete than most other messages.
+///
+/// However, the `O(1)` operations come at a cost of higher constant factor than the `pure/Queue` implementation:
+/// - 'pop' operations are on average 4x more expensive
+/// - 'push' operations are on average 10x more expensive
+///
+/// For better performance across multiple operations and when the spikes in single operations are not a problem, use `pure/Queue`.
+/// For guaranteed `O(1)` operations, use `pure/RealTimeQueue`.
+///
+/// ---
+///
+/// The interface is purely functional, not imperative, and queues are immutable values.
+/// In particular, Queue operations such as push and pop do not update their input queue but, instead, return the
+/// value of the modified Queue, alongside any other data.
+/// The input queue is left unchanged.
+///
+/// Examples of use-cases:
+/// - Queue (FIFO) by using `pushBack()` and `popFront()`.
+/// - Stack (LIFO) by using `pushFront()` and `popFront()`.
+/// - Deque (double-ended queue) by using any combination of push/pop operations on either end.
+///
+/// A Queue is internally implemented as a real-time double-ended queue based on the paper
+/// "Real-Time Double-Ended Queue Verified (Proof Pearl)". The implementation maintains
+/// worst-case constant time `O(1)` for push/pop operations through gradual rebalancing steps.
+///
+/// Construction: Create a new queue with the `empty<T>()` function.
+///
+/// Note that some operations that traverse the elements of the queue (e.g. `forEach`, `values`) preserve the order of the elements,
+/// whereas others (e.g. `map`, `contains`, `any`) do NOT guarantee that the elements are visited in any order.
```

Comment:
```suggestion
/// whereas others (e.g. `map`, `contains`, `any`) do NOT guarantee that the elements are visited in any order.
```
I think `any` does now, does it not?

Reply:
Good catch, I have to update that list now, and I'll put a similar note on the pure/Queue for now.